### ISSUE TITLE
unused-variable false positive

### ISSUE THREAD

**Issue by douglas-raillard-arm** [PyCQA/pylint#4020]
### Steps to reproduce
This seems to be the minimal reproducer, 
```
def f():
    def g():
        return type

    class C(metaclass=g()):
        pass

    return C
```

### Current behavior
```
test.py:5:4: W0612: Unused variable 'g' (unused-variable)
```

### Expected behavior
No error, since the function is used.

### pylint --version output
```
pylint 2.6.0
astroid 2.4.2
Python 3.8.7 (default, Dec 24 2020, 17:53:09) 
[GCC 10.2.0]
```

**Comment by hippo91** [comment]
@douglas-raillard-arm thanks for the report. I can reproduce it.

**Comment by sudomyju** [comment]
Hi guys, 

same applies for cycle/s (I guess)
'''def ster(oids):
    for i in range(oids)
        print("oids")
'''

while in the cycle (it is kind of disputable) if it is really unused when it runs the cycle, I think the better pylint result in this case would be either some "info" like "Consider using _ when not planning to utilize the variable used for constructing the cycle".

(On other hand, I don't understand why are the real problems/errors are mixed up with some sort of highly opinioned assumptions on how the code should look like of pylint's which are considered errors as well.
where the prime example of this is sort of error is being considering an empty line a triling space!)


'''
pylint 2.6.0
astroid 2.4.2
Python 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:06:47) [MSC v.1914 32 bit (Intel)]'''

Hope this helps,
Su
